src/common/node.rs

```diff
@@ -10,9 +10,35 @@ pub trait Node {
     fn height(&self) -> u32;
     fn leaf_key(&self) -> Self::Key;
     fn is_leaf(&self) -> bool;
+    fn is_node(&self) -> bool;
 }
 
-pub trait ParentNode: Node {
-    fn left_child(&self) -> Self;
-    fn right_child(&self) -> Self;
+pub trait ParentNode: Node
+where
+    Self: Sized,
+{
+    type Error;
+
+    fn left_child(&self) -> ChildResult<Self>;
+    fn right_child(&self) -> ChildResult<Self>;
+}
+
+#[allow(type_alias_bounds)]
+pub type ChildResult<T: ParentNode> = Result<T, ChildError<T::Key, T::Error>>;
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+pub enum ChildError<Key, E> {
+    #[cfg_attr(feature = "std", error("Child with key {0} was not found in storage"))]
+    ChildNotFound(Key),
```

Comment:
I decided to make the case where the child is not found return an error, instead of returning `None`. Since nodes are only used internally in tree traversal (and not arbitrarily called by client code), when a child is not found, it means that there is either data corruption or an error in the algorithm. These are fatal cases, and tree traversal must exit with an error.